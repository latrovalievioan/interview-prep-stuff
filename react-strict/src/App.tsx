import { useState, useEffect } from "react";
import "./App.css";

const BadTimer = () => {
  const [time, setTime] = useState(0);

  useEffect(() => {
    setInterval(() => {
      setTime((time) => time + 0.1);
    }, 100);
  }, []);

  return (
    <div>
      <h2>
        Bad Timer: <br /> {time.toFixed(2)}
      </h2>
    </div>
  );
};

const GoodTimer = () => {
  const [time, setTime] = useState(0);

  useEffect(() => {
    const interval = setInterval(() => {
      setTime((time) => time + 0.1);
    }, 100);

    return () => clearInterval(interval);
  }, []);

  return (
    <div>
      <h2>
        Good Timer: <br /> {time.toFixed(2)}
      </h2>
    </div>
  );
};

function App() {
  return (
    <div>
      <BadTimer />
      <GoodTimer />
    </div>
  );
}

export default App;
