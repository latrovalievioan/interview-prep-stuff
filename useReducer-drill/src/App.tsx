import { useReducer } from 'react';

type State = {
  count: number;
  count2: number;
};

type ActionTypes =
  | {
      type: 'setCount';
      payload: number;
    }
  | {
      type: 'setCount2';
      payload: number;
    };

const reducer = (state: State, action: ActionTypes) => {
  switch (action.type) {
    case 'setCount':
      return { ...state, count: action.payload };
    case 'setCount2':
      return { ...state, count2: action.payload };
  }
};

const initialState: State = {
  count: 0,
  count2: 0,
};

function App() {
  const [state, dispatch] = useReducer(reducer, initialState);
  const dispatchIncrementCountByOne = () =>
    dispatch({ type: 'setCount', payload: state.count + 1 });
  const dispatchDecrementCountByOne = () =>
    dispatch({ type: 'setCount', payload: state.count - 1 });

  const dispatchIncrementCount2ByOne = () =>
    dispatch({ type: 'setCount2', payload: state.count2 + 1 });
  const dispatchDecrementCount2ByOne = () =>
    dispatch({ type: 'setCount2', payload: state.count2 - 1 });

  return (
    <>
      <div
        style={{
          display: 'flex',
          flexDirection: 'column',
          alignItems: 'center',
          justifyContent: 'center',
          border: '1px solid white',
        }}
      >
        <h1 data-testid="count display">Count: {state.count}</h1>
        <div style={{ display: 'flex' }}>
          <button onClick={dispatchIncrementCountByOne}>Increment Count</button>
          <button onClick={dispatchDecrementCountByOne}>Decrement Count</button>
        </div>
      </div>
      <div
        style={{
          display: 'flex',
          flexDirection: 'column',
          alignItems: 'center',
          justifyContent: 'center',
          border: '1px solid white',
        }}
      >
        <h1 data-testid="count2 display">Count2: {state.count2}</h1>
        <div style={{ display: 'flex' }}>
          <button onClick={dispatchIncrementCount2ByOne}>
            Increment Count2
          </button>
          <button onClick={dispatchDecrementCount2ByOne}>
            Decrement Count2
          </button>
        </div>
      </div>
    </>
  );
}

export default App;
