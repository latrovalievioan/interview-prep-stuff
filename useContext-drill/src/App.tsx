import { Dispatch, SetStateAction, createContext, useState } from 'react';
import './App.css';
import { Controls } from './Controls';
import { Display } from './Display';

export type Store = {
  text: string;
  setText: Dispatch<SetStateAction<string>>;
};

const defaultStore: Store = {
  text: '',
  setText: () => null,
};

export const Context = createContext(defaultStore);

function App() {
  const [text, setText] = useState<string>('');

  const storeValue = {
    text,
    setText,
  };

  return (
    <Context.Provider value={storeValue}>
      <Controls />
      <Display />
    </Context.Provider>
  );
}

export default App;
