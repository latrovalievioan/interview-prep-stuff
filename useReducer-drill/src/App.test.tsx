import App from './App';
import { describe, test, expect, afterEach, beforeEach } from 'vitest';
import { render, act, cleanup, screen } from '@testing-library/react';

describe(App, () => {
  const { getByTestId, getByText } = screen;

  test('Renders correctly', () => {
    const app = render(<App />);
    expect(app).toMatchSnapshot();
  });

  beforeEach(() => {
    render(<App />);
  });

  afterEach(cleanup);

  test('Counts start with a correct initial value', () => {
    const countValue = getByTestId('count display').textContent;
    const count2Value = getByTestId('count2 display').textContent;

    expect([countValue, count2Value]).toEqual(['Count: 0', 'Count2: 0']);
  });

  test('First count is incrementing correctly', () => {
    const countValue = getByTestId('count display');
    expect(countValue.textContent).toEqual('Count: 0');
    const countButton = getByText('Increment Count');
    act(() => countButton.click());
    expect(countValue.textContent).toEqual('Count: 1');
    act(() => countButton.click());
    expect(countValue.textContent).toEqual('Count: 2');
  });

  test('First count is decremented correctly', () => {
    const countValue = getByTestId('count display');
    expect(countValue.textContent).toEqual('Count: 0');
    const countButton = getByText('Decrement Count');
    act(() => countButton.click());
    expect(countValue.textContent).toEqual('Count: -1');
    act(() => countButton.click());
    expect(countValue.textContent).toEqual('Count: -2');
  });

  test('Second count2 is incrementing correctly', () => {
    const count2Value = getByTestId('count2 display');
    expect(count2Value.textContent).toEqual('Count2: 0');
    const count2Button = getByText('Increment Count2');
    act(() => count2Button.click());
    expect(count2Value.textContent).toEqual('Count2: 1');
    act(() => count2Button.click());
    expect(count2Value.textContent).toEqual('Count2: 2');
  });

  test('Second count2 is decremented correctly', () => {
    const count2Value = getByTestId('count2 display');
    expect(count2Value.textContent).toEqual('Count2: 0');
    const count2Button = getByText('Decrement Count2');
    act(() => count2Button.click());
    expect(count2Value.textContent).toEqual('Count2: -1');
    act(() => count2Button.click());
    expect(count2Value.textContent).toEqual('Count2: -2');
  });
});
