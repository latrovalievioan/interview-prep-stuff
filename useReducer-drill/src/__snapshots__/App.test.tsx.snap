// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > Renders 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="display: flex; flex-direction: column; align-items: center; justify-content: center; border: 1px solid white;"
      >
        <h1
          data-testid="count display"
        >
          Count: 
          0
        </h1>
        <div
          style="display: flex;"
        >
          <button>
            Increment Count
          </button>
          <button>
            Decrement Count
          </button>
        </div>
      </div>
      <div
        style="display: flex; flex-direction: column; align-items: center; justify-content: center; border: 1px solid white;"
      >
        <h1
          data-testid="count2 display"
        >
          Count2: 
          0
        </h1>
        <div
          style="display: flex;"
        >
          <button>
            Increment Count2
          </button>
          <button>
            Decrement Count2
          </button>
        </div>
      </div>
    </div>
    <div>
      <div
        style="display: flex; flex-direction: column; align-items: center; justify-content: center; border: 1px solid white;"
      >
        <h1
          data-testid="count display"
        >
          Count: 
          0
        </h1>
        <div
          style="display: flex;"
        >
          <button>
            Increment Count
          </button>
          <button>
            Decrement Count
          </button>
        </div>
      </div>
      <div
        style="display: flex; flex-direction: column; align-items: center; justify-content: center; border: 1px solid white;"
      >
        <h1
          data-testid="count2 display"
        >
          Count2: 
          0
        </h1>
        <div
          style="display: flex;"
        >
          <button>
            Increment Count2
          </button>
          <button>
            Decrement Count2
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="display: flex; flex-direction: column; align-items: center; justify-content: center; border: 1px solid white;"
    >
      <h1
        data-testid="count display"
      >
        Count: 
        0
      </h1>
      <div
        style="display: flex;"
      >
        <button>
          Increment Count
        </button>
        <button>
          Decrement Count
        </button>
      </div>
    </div>
    <div
      style="display: flex; flex-direction: column; align-items: center; justify-content: center; border: 1px solid white;"
    >
      <h1
        data-testid="count2 display"
      >
        Count2: 
        0
      </h1>
      <div
        style="display: flex;"
      >
        <button>
          Increment Count2
        </button>
        <button>
          Decrement Count2
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`App > Renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="display: flex; flex-direction: column; align-items: center; justify-content: center; border: 1px solid white;"
      >
        <h1
          data-testid="count display"
        >
          Count: 
          0
        </h1>
        <div
          style="display: flex;"
        >
          <button>
            Increment Count
          </button>
          <button>
            Decrement Count
          </button>
        </div>
      </div>
      <div
        style="display: flex; flex-direction: column; align-items: center; justify-content: center; border: 1px solid white;"
      >
        <h1
          data-testid="count2 display"
        >
          Count2: 
          0
        </h1>
        <div
          style="display: flex;"
        >
          <button>
            Increment Count2
          </button>
          <button>
            Decrement Count2
          </button>
        </div>
      </div>
    </div>
    <div>
      <div
        style="display: flex; flex-direction: column; align-items: center; justify-content: center; border: 1px solid white;"
      >
        <h1
          data-testid="count display"
        >
          Count: 
          0
        </h1>
        <div
          style="display: flex;"
        >
          <button>
            Increment Count
          </button>
          <button>
            Decrement Count
          </button>
        </div>
      </div>
      <div
        style="display: flex; flex-direction: column; align-items: center; justify-content: center; border: 1px solid white;"
      >
        <h1
          data-testid="count2 display"
        >
          Count2: 
          0
        </h1>
        <div
          style="display: flex;"
        >
          <button>
            Increment Count2
          </button>
          <button>
            Decrement Count2
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="display: flex; flex-direction: column; align-items: center; justify-content: center; border: 1px solid white;"
    >
      <h1
        data-testid="count display"
      >
        Count: 
        0
      </h1>
      <div
        style="display: flex;"
      >
        <button>
          Increment Count
        </button>
        <button>
          Decrement Count
        </button>
      </div>
    </div>
    <div
      style="display: flex; flex-direction: column; align-items: center; justify-content: center; border: 1px solid white;"
    >
      <h1
        data-testid="count2 display"
      >
        Count2: 
        0
      </h1>
      <div
        style="display: flex;"
      >
        <button>
          Increment Count2
        </button>
        <button>
          Decrement Count2
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`App > Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="display: flex; flex-direction: column; align-items: center; justify-content: center; border: 1px solid white;"
      >
        <h1
          data-testid="count display"
        >
          Count: 
          0
        </h1>
        <div
          style="display: flex;"
        >
          <button>
            Increment Count
          </button>
          <button>
            Decrement Count
          </button>
        </div>
      </div>
      <div
        style="display: flex; flex-direction: column; align-items: center; justify-content: center; border: 1px solid white;"
      >
        <h1
          data-testid="count2 display"
        >
          Count2: 
          0
        </h1>
        <div
          style="display: flex;"
        >
          <button>
            Increment Count2
          </button>
          <button>
            Decrement Count2
          </button>
        </div>
      </div>
    </div>
    <div>
      <div
        style="display: flex; flex-direction: column; align-items: center; justify-content: center; border: 1px solid white;"
      >
        <h1
          data-testid="count display"
        >
          Count: 
          0
        </h1>
        <div
          style="display: flex;"
        >
          <button>
            Increment Count
          </button>
          <button>
            Decrement Count
          </button>
        </div>
      </div>
      <div
        style="display: flex; flex-direction: column; align-items: center; justify-content: center; border: 1px solid white;"
      >
        <h1
          data-testid="count2 display"
        >
          Count2: 
          0
        </h1>
        <div
          style="display: flex;"
        >
          <button>
            Increment Count2
          </button>
          <button>
            Decrement Count2
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="display: flex; flex-direction: column; align-items: center; justify-content: center; border: 1px solid white;"
    >
      <h1
        data-testid="count display"
      >
        Count: 
        0
      </h1>
      <div
        style="display: flex;"
      >
        <button>
          Increment Count
        </button>
        <button>
          Decrement Count
        </button>
      </div>
    </div>
    <div
      style="display: flex; flex-direction: column; align-items: center; justify-content: center; border: 1px solid white;"
    >
      <h1
        data-testid="count2 display"
      >
        Count2: 
        0
      </h1>
      <div
        style="display: flex;"
      >
        <button>
          Increment Count2
        </button>
        <button>
          Decrement Count2
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
